crates/sound/src/stream.rs

```
@@ -234,29 +234,46 @@ impl Default for PcmParams {
 }
 
 pub struct Buffer {
-    pub bytes: Vec<u8>,
+    // TODO: to make private and add len usize
+    pub data_descriptor: virtio_queue::Descriptor,
     pub pos: usize,
     pub message: Arc<IOMessage>,
 }
 
 impl std::fmt::Debug for Buffer {
     fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
         fmt.debug_struct(stringify!(Buffer))
-            .field("bytes", &self.bytes.len())
             .field("pos", &self.pos)
             .field("message", &Arc::as_ptr(&self.message))
             .finish()
     }
 }
 
 impl Buffer {
-    pub fn new(bytes: Vec<u8>, message: Arc<IOMessage>) -> Self {
+    pub fn new(data_descriptor: virtio_queue::Descriptor, message: Arc<IOMessage>) -> Self {
         Self {
-            bytes,
             pos: 0,
+            data_descriptor,
             message,
         }
     }
+
+    pub fn consume(&self, buf: &mut [u8]) -> u32 {
+        let addr = self.data_descriptor.addr();
+        let offset = self.pos as u64;
+        let len = self
+            .message
+            .desc_chain
+            .memory()
+            .read(
+                buf,
+                addr.checked_add(offset)
+                    .expect("invalid guest memory address"),
+            )
+            .map_err(|_| StreamError::DescriptorReadFailed)
+            .unwrap();
```

Should be better to propagate the error?